[Emotion] Convert `EuiFacetButton` and `EuiFacetGroup`

miukimiu opened the pull request:
### Summary

Provide a detailed summary of your PR. Explain how you arrived at your solution. If it includes changes to UI elements include a screenshot or gif.

### Checklist

- [ ] Checked in both **light and dark** modes
- [ ] Checked in **mobile**
- [ ] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**
- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples
- [ ] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- [ ] Checked for **breaking changes** and labeled appropriately
- [ ] Checked for **accessibility** including keyboard-only and screenreader modes
- [ ] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart
- [ ] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented on src/components/facet/facet_button.tsx:
```diff
@@ -97,25 +113,29 @@ export const EuiFacetButton: FunctionComponent<EuiFacetButtonProps> = ({
+
             <span
+              css={cssTextStyles}
               className="euiFacetButton__text"
```
In use in Kibana.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

cchaos commented on src/components/facet/facet_button.tsx:
```diff
@@ -97,25 +113,29 @@ export const EuiFacetButton: FunctionComponent<EuiFacetButtonProps> = ({
   return (
     <EuiInnerText>
       {(ref, innerText) => (
         <button
+          css={cssStyles}
```
Is it not possible just to use the EuiButtonDisplay component here instead of trying to re-create all the styles through mixins?

cchaos commented on src/components/button/button_content.styles.ts:
```diff
@@ -0,0 +1,55 @@
+// </button>
+// 1. Apply margin to all but last item in the flex.
+// 2. Margin gets flipped because of the row-reverse.
+export const euiButtonContentCSS = (
```
:sigh: I was really hoping we could get away from creating component-based mixins that other components use. I'd rather see re-usable content-style components instead.

cchaos commented on src/components/button/button_content.styles.ts:
```diff
@@ -0,0 +1,55 @@
+  if (isReverse) {
+    cssStyles = `flex-direction: row-reverse;
+  
+      > * + * {
```
If we must use mixins, we need to figure out a way to not use universal selectors (`*`). These are usually how we get ourselves into trouble if we ever manipulate the rendered dom structure. This looks like it's adding margin that could just be using `gap` here instead?

cchaos commented on src/components/button/button_content.styles.ts:
```diff
@@ -0,0 +1,55 @@
+      vertical-align: middle;
+      ${cssStyles};
+  
+      [class*='euiButtonContent__icon'],
+      [class*='euiButtonContent__spinner'] {
```
These should be applied only where this class is created. No more targeting very specific classes in generic mixins.

cchaos commented on src-docs/src/views/facet/facet_layout.js:
```diff
@@ -146,22 +144,8 @@ export default () => {
-      <EuiSpacer />
-
-      <EuiTitle size="s">
-        <h3>Horizontal and large gutter</h3>
```
Good catch

cchaos commented on src/components/facet/__snapshots__/facet_group.test.tsx.snap:
```diff
@@ -3,43 +3,43 @@
 exports[`EuiFacetGroup props layout horizontal is rendered 1`] = `
 <div
-  class="euiFlexGroup euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap euiFacetGroup euiFacetGroup--horizontal euiFacetGroup--gutterMedium"
+  class="euiFlexGroup euiFlexGroup--directionRow euiFlexGroup--responsive euiFlexGroup--wrap euiFacetGroup css-1qfxtqy-euiFacetGroup-m"
```
Looks like we're missing the output in the classname for `vertical` vs `horizontal`.

cchaos commented on src/components/facet/facet_group.styles.ts:
```diff
@@ -0,0 +1,76 @@
+    ${layoutStyles}
+
+    > [class*='euiFacetButton'] {
+      ${logicalCSS('margin-vertical', buttonMarginVerticalSize)};
```
Do we need this class target or will `gap` take care of it?

cchaos commented on src/components/facet/facet_group.tsx:
```diff
@@ -69,6 +55,7 @@ export const EuiFacetGroup: FunctionComponent<EuiFacetGroupProps> = ({
       direction={direction}
       wrap={wrap}
       gutterSize="none"
+      css={cssStyles}
```
It seems like we're overrdiing a lot of innate properties that the FlexGroup gives us. I would either pass through the `gutterSize` here or change this to a basic div.

cchaos commented on src/components/facet/facet_group.styles.ts:
```diff
@@ -0,0 +1,76 @@
+  const layoutStyles =
+    layout === 'horizontal'
+      ? `
+      display: flex;
+      flex-direction: row;
```
I know this was probably a direct 1:1 convertions, but there is a lot of repetition per gutter size and why we're missing the `vertical` vs `horizontal` keys in the output. My suggestion is to render the common properties as static keys and only use this function for changing the gap size. I think this could be more succinct.

cchaos commented on src/components/facet/facet_group.styles.ts:
```diff
@@ -0,0 +1,76 @@
+      flex-direction: row;
+      gap: 0 ${gutterSize};`
+      : `
+      > [class*='euiFacetButton'] {
+        width: 100%;
+      }
```
I checked the render and I don't think this is needed.

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,108 @@
+    flex-grow: 1;
+    vertical-align: middle;
+  `,
+  euiFacetButton__spinner: css``,
+  euiFacetButton__quantity: css``,
```
Without any styles being created, do we need these keys?

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,108 @@
+        cursor: not-allowed;
+      }
+
+      [class*='euiFacetButton__icon'],
+      [class*='euiFacetButton__quantity'] {
```
I would pass `disabled` to the icon and quantity elements. Reminder that wherever possible, we should be passing / creating styles based on prop values directly on those elements and not targeting class names

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,108 @@
+    text-decoration: none;
+
+    ${euiCanAnimate} {
+      transition: all ${euiTheme.animation.fast} ease-in;
```
Let's try to remove replace `all` properties with targeted ones since not all properties have good performance when animated.

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,108 @@
+    ${euiButtonBaseCSS(_theme)}
+    ${euiFontSize('s', euiTheme)}
+    ${logicalCSS('height', euiTheme.size.xl)};
+    text-align: left;
```
There's a logical property for tex alignment too `logicalTextAlignCSS('left')`

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    &:focus:not(:disabled) [class*='euiFacetButton__text'] {
+      ${euiLinkFocusCSS(_theme)}
```
Another reason why we should just be using a basic button component so that we don't have to constantly redeclare focus states.

miukimiu replied:
This is a special focus state different from the buttons. It adds a 
 `text-decoration-thickness: 2px;`. 🤷🏽‍♀️

cchaos reviewed (changes requested):
The straight conversion works well, but I think we could really clean up all the manual button styles

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented on src/components/button/button_display/button_display.styles.ts:
```diff
@@ -0,0 +1,65 @@
+
+// Provides a solid reset and base for handling sizing layout
+// Does not include any visual styles
+export const euiButtonBaseCSS = () => {
```
Because the **EuiButtonDisplay** is now more flexible, probably we don't need to export this.

miukimiu commented on src/components/button/button_display/button_display.tsx:
```diff
@@ -0,0 +1,99 @@
+
+export type ButtonSize = 's' | 'm';
+
+export type EuiButtonDisplayProps = Omit<EuiButtonProps, 'color'> &
```
Because we don't need a color logic for the `EuiFacetButton` I'm omitting this. The idea is then we start converting the **EuiButton** we rethink the logic for applying color with Emotion.

cchaos replied:
What's the intention with this component vs the current `EuiButtonDisplay`? They're both named exactly the same so I wonder if there will be export name conflicts. Should we deprecate the old one by renaming it? Or if this is truly for internal only, name it with an underscore or possibly as a namespaced component like `EuiButton.Display`.

I'm all for creating a more generic re-usable button type, we just need to be absolutely clear how this should be used vs the current one.

miukimiu replied:
I wanted to use the `EuiButtonDisplay` but I found it very strict. So instead of starting to modify the current `EuiButtonDisplay` and probably break things in other buttons that depend on it, I decided to create a new one. The idea is to then deprecate the old one. Some reasons why I took this decision (https://github.com/elastic/eui/pull/5878#issuecomment-1132935586).

- The way I see this component is for internal use only so I'll try the namespaced component approach.
- When we convert all the buttons we deprecate the old one. For now, the new `EuiButtonDisplay` is very basic but needs more props to handle color (for EuiButton). And these new props we should only add once we need them when we start the other button conversions.

cchaos replied:
Yeah, the point of deprecations though is to show people **now** that something is going to get removed. I'd rename the old one now and comment with an action as to how to update to the new version. Then once all the usages have been updated it can be removed.

miukimiu commented on src/test/internal/render_custom_styles.tsx:
```diff
@@ -37,7 +37,7 @@ export const shouldRenderCustomStyles = (component: ReactElement) => {
     expect(componentNode).toHaveLength(1);
     // css
     expect(componentNode.attr('class')).toEqual(
-      expect.stringMatching(/css-[\d\w]{6,7}-css/) // should have generated an emotion class ending with -css
+      expect.stringMatching(/css-[\d\w]{6,7}(-[a-zA-Z0-9]+)*-css/) // should have generated an emotion class ending with -css
```
@constancecchen this is meant to fix the same issue I run in https://github.com/elastic/eui/pull/5692#issuecomment-1115989431. But because #5692 is not merged yet, I just copied and paste what Chandler did there.

But I saw you had some suggestions https://github.com/elastic/eui/pull/5692#commitcomment-72872938. Should I add your suggestion instead?

constancecchen replied:
Yes, please use my suggestion instead - it's a shorter way of writing Chandler's regex :)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented on src/components/button/button_display/button_display_content.tsx:
```diff
@@ -0,0 +1,97 @@
+  return (
+    <span {...contentProps} css={cssStyles}>
+      {icon}
+      {text && <span {...textProps}>{text}</span>}
+      {children}
```
The text is basically the `children` from `src/components/button/button_content.tsx`. The children is always `text` and it's always wrapped in a span that can receive ` {...textProps}`. So this makes the prop less misleading. This way we know what we should pass.  
Having a `children` gives us the possibility to pass custom components. That we don't want to be wrapped in a span. 

The way the `isLoading` work is to replace the current icon. That's why we can only have one icon and pick what is the `iconSide`. Just because I didn't want to complicate I just decided to support custom components. Where we can just define our custom layout and our logic for loading states.

cchaos replied:
I think the struggle I"m having is that this prioritizes `children` over `text` making it harder to simply pass the button label as the child. The more common case is just displaying `text` not custom components. Why can't `children` always be wrapped in the span whether it's just plain text or not?

cchaos replied:
Or what if we evaluated `children`'s type and if it's simply a `string`, we wrap in the span, otherwise it doesn't get wrapped?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented on src/components/button/button_display/_button_display_content.tsx:
```diff
@@ -0,0 +1,93 @@
+    );
+  }
+
+  const isText = typeof children === 'string';
+
+  return (
+    <span {...contentProps} css={cssStyles}>
+      {icon}
+      {isText ? <span {...textProps}>{children}</span> : children}
+    </span>
+  );
```
> Or what if we evaluated children's type and if it's simply a string, we wrap in the span, otherwise it doesn't get wrapped?

@cchaos I implemented your suggestion.

miukimiu commented on src/components/button/button_display/_button_display_content.styles.ts:
```diff
@@ -0,0 +1,43 @@
+    pointer-events: auto;
+    cursor: not-allowed;
+
+    [class*='euiButtonDisplayContent__spinner'] {
+      ${euiLoadingSpinnerBorderColorsCSS(euiTheme, 'currentColor')}
+    }
```
> If this new component does though render this spinner, we should be figuring out how to make this work otherwise it will not render correctly. My guess is that we should just allow passing a custom color to the EuiLoadingSpinner component. But I'd do that in a separate PR (to merge before this one).

@cchaos I added a mixin (strict conversion for now) but currently the **EuiFacetButton** doesn't need this mixin and the facet spinner doesn't live in `.euiButtonDisplayContent__spinner` so it doesn't get this style. 

The border color seems to be changed in the `EuiButton` to make the gray borders darker:

<img width="321" alt="button__@2x" src="https://user-images.githubusercontent.com/2750668/170492157-c662746a-f5be-4f0a-9f05-7f435c8e5801.png">

So it doesn't make sense for buttons without backgrounds. But I leave this strict conversion. 

Also, all the other buttons are using the **deprecated** button display and button content. So adding this in practice doesn't do anything but I think for now it's ok to leave all the styles converted.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented on src/components/button/button.tsx:
```diff
@@ -174,7 +174,8 @@ export const EuiButton: FunctionComponent<Props> = ({
   return (
-    <EuiButtonDisplay
+    // eslint-disable-next-line react/jsx-pascal-case
+    <EuiButtonDisplay_Deprecated
```
@thompsongl @constancecchen,
 
Because this component lives in the same file of `button.tsx` I can't `import { EuiButtonDisplay_Deprecated as EuiButtonDisplay }`. So I'm not sure if it is a good idea to rename it.

I tried to move the `EuiButtonDisplay` to a different file but because it uses some `EuiButton`  props I run into some circular dependencies issues. Then I moved all the `EuiButton` props to the new button display file (https://github.com/elastic/eui/pull/5878/commits/9c669da2ba46541f153a3cd264613ac330e70f85). But I reverted because I was already changing too much... 

So I need some advise or help here on what's the best approach to deprecate this component. Just rename it... Leave it as it is with a comment. Move the component to another file including the `EuiButton` props... 🤷🏽‍♀️

thompsongl replied:
Ok maybe rename it `EuiButtonDisplayDeprecated` and update all the places it's used to that name. It's internal use only so we're not concerned about consumers mistakenly using the wrong one.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented:
@cchaos the PR is ready for review.

I have some questions about the deprecation of the `EuiButtonDisplay` and `EuiButtonContent`. See https://github.com/elastic/eui/pull/5878#discussion_r882689312.

miukimiu commented on src/components/button/button.tsx:
```diff
@@ -295,4 +302,4 @@ export const EuiButtonDisplay = forwardRef<HTMLElement, EuiButtonDisplayProps>(
   }
 );
-EuiButtonDisplay.displayName = 'EuiButtonDisplay';
+EuiButtonDisplay_Deprecated.displayName = 'EuiButtonDisplay';
```
I'm displaying a different name so that snapshots don't end up with `<EuiButtonDisplay_Deprecated />`

miukimiu replied:
@thompsongl another concern of renaming `EuiButtonDisplay` to `EuiButtonDisplay_Deprecated` was because some components that use a `EuiButton` were showing this `EuiButtonDisplay_Deprecated` on the snapshots. But during my 1-to-1 with @chandlerprall he recommended changing the display name. And the snapshots no longer show `EuiButtonDisplay_Deprecated`.

cchaos commented on src/components/button/button_display/_button_display.tsx:
```diff
@@ -0,0 +1,209 @@
+>;
+
+/**
+ * The `color` is being ommited because the first component using this new component EuiFacetButton doesn't requiere a background color.
+ * But once we start using this component in buttons that can have a background we will require the color.
```
This button **display** component shouldn't ever really contain any visual styles other than how to display the layout. I would remove all references to `color` completely.

cchaos replied:
We can always add it back in later if we truly find it necessary.

cchaos commented on src/components/facet/facet_button.styles.ts:
```diff
@@ -0,0 +1,79 @@
+      }
+    }
+  `,
+  // Selections
+  isSelected: css``,
+  unSelected: css``,
```
If these don't provide any styling, is it necessary to have the keys?

cchaos commented on src/components/loading/loading_spinner.styles.ts:
```diff
@@ -35,6 +35,19 @@ const spinnerColors = (main: string, highlight: string) => {
   return `${highlight} ${main} ${main} ${main}`;
 };
 
+export const euiLoadingSpinnerBorderColorsCSS = (
```
This is almost a complete duplicate of the above `spinnerColors`. If you don't go the route of allowing customization through props, can you consolidate these two?

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

miukimiu commented:
@cchaos I went through your review:

- Deleted all the types that were not being used in the new  **EuiButtonDisplay**. 
- Added a `color` prop to **EuiLoadingSpinner**

@thompsongl I renamed both deprecated components to `EuiButtonDisplayDeprecated` and `EuiButtonContentDeprecated`. Nore more `_Deprecated` on the names.

I updated the PR description to add the new changes.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

cchaos commented on src/components/loading/loading_spinner.tsx:
```diff
@@ -16,19 +16,31 @@ import { euiLoadingSpinnerStyles } from './loading_spinner.styles';
 export type EuiLoadingSpinnerSize = typeof SIZES[number];
 
+export type EuiLoadingSpinnerColor = {
+  border?: string;
+  highlight?: string;
```
```suggestion
  border?: CSSProperties['border-color'];
  highlight?: CSSProperties['border-color'];
```

I know it evaluates to the same thing (string), but it makes it clearer the intention of where the color will be applied.

cchaos commented on upcoming_changelogs/5878.md:
```diff
@@ -0,0 +1,4 @@
+**CSS-in-JS**
+
+- Converted `EuiFacetGroup` and `EuiFacetButton` to Emotion
```
```suggestion
- Converted `EuiFacetGroup` and `EuiFacetButton` to Emotion and removed `$euiFacetGutterSizes`
```

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5878/
